### PR TITLE
[Model] Cosmos3: Reduce video output transfer overhead

### PR DESCRIPTION
**Pull request by rahul-steiger-nv**
## Purpose

Reduce Cosmos3 video response latency and peak GPU memory after VAE decode.

Previously, Cosmos3 returned decoded video as a bfloat16 tensor in
`[B, C, T, H, W]` layout. The diffusion IPC path then promoted bfloat16 to
float32 for NumPy/shared-memory compatibility, and postprocessing converted the
result to display range before the video encoder converted it again to uint8.
For a 1280x720, 189-frame video, this widened the shared-memory payload to about
2.09 GB even though the final encoder input is only about 0.52 GB of uint8
frames.

This change converts decoded video on the output rank directly to display-ready
uint8 `[B, T, H, W, C]` before IPC. The existing asynchronous pinned-memory D2H
path then transfers the narrower payload without an additional format
conversion.

Key details:

- A 64 MiB temporal chunk budget bounds conversion intermediates at the
  post-VAE memory peak.
- Historical bfloat16 arithmetic and final display bytes are preserved.
- Conversion runs only on the output rank and is skipped for T2I and CPU
  offload.
- Guardrails consume uint8 directly, avoiding a float round trip.
- `np`, `pt`, and `pil` output types remain supported.

## Test Plan

- CPU tests for output parity, chunking, rank/offload gates, output types, and
  guardrails.
- CUDA regression for sub-threshold device output reaching NumPy postprocess.
- Online-serving T2I/T2V/I2V/V2V smoke coverage.
- Same-node before/after serving benchmarks.

```bash
pytest -q \
  tests/diffusion/models/cosmos3/test_cosmos3_pipeline.py \
  tests/diffusion/models/cosmos3/test_cosmos3_guardrails.py
```

**vLLM Version:** 0.25.0

**vLLM-Omni Baseline Commit:** `a4ea67a21b`

**vLLM-Omni Optimized Commit:** `d80fa73a8a`

### Unit and E2E correctness

- CPU: `89 passed, 1 skipped`; CUDA regression: `1 passed`; online smoke:
  `2 passed`; Ruff: all checks passed.
- Every baseline and optimized benchmark completed 5/5 measured requests.

### Single-GPU benchmark

NVIDIA GB300; T2V 1280x720, 189 frames, 35 steps; no VAE tiling;
concurrency 1; 1 warmup + 5 measured requests; CUDNN attention; guardrails off.

| revision | mean latency | median latency | throughput | peak GPU memory | mean stage generation |
| --- | ---: | ---: | ---: | ---: | ---: |
| baseline | 87.243 s | 87.295 s | 0.01146 req/s | 49.572 GB | 83.010 s |
| optimized | 84.131 s | 84.132 s | 0.01189 req/s | 47.560 GB | 82.144 s |
| change | **-3.112 s (-3.6%)** | **-3.163 s (-3.6%)** | **+0.00042 req/s (+3.7%)** | **-2.012 GB (-4.1%)** | **-0.865 s (-1.0%)** |

### Four-GPU benchmark

4x NVIDIA GB300; same workload; CFG parallel 2; Ulysses 2; VAE patch
parallel 4 with `spatial_shard_width`; concurrency 1; 1 warmup + 5 measured
requests; CUDNN attention; guardrails off.

| revision | mean latency | median latency | throughput | peak GPU memory | mean stage generation |
| --- | ---: | ---: | ---: | ---: | ---: |
| baseline | 28.846 s | 28.712 s | 0.03467 req/s | 41.240 GB | 24.622 s |
| optimized | 27.294 s | 27.285 s | 0.03664 req/s | 39.244 GB | 23.631 s |
| change | **-1.552 s (-5.4%)** | **-1.427 s (-5.0%)** | **+0.00197 req/s (+5.7%)** | **-1.996 GB (-4.8%)** | **-0.991 s (-4.0%)** |

### Notes

Pipeline `forward` and VAE decode times were unchanged, confirming that the gain
comes from output handling. MP4 encoding improved by 0.544 s (24.9%) on one GPU
and 0.547 s (25.1%) on four GPUs. The four-GPU stage-generation reduction
(0.991 s) plus encoding reduction accounts for nearly all of its 1.552 s E2E
improvement. Byte-parity tests confirm chunking does not alter display output.

### CODE REVIEW

**Comment by rahul-steiger-nv** [comment]
Friendly ping for review when you have a chance 🙂 Would appreciate any feedback. Thanks!

**Comment by rahul-steiger-nv** [comment]
I will leave this PR open until #6477 and #6541 have been resolved. However, I do not plan on merging this into the main branch

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 1]
```diff
@@ -688,7 +750,20 @@ def _postprocess_action(action: Any, metadata: dict[str, Any]) -> Any:
         guardrails_enabled = is_guardrails_enabled(od_config, sampling_params)
         if guardrails_enabled:
             video = check_video_safety(video)
-        processed_video = video_processor.postprocess_video(video, output_type=output_type)
+        if video.dtype == torch.uint8:
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:836]: np and pt output types silently change dtype, layout, and value range

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 2]
```diff
@@ -688,7 +750,20 @@ def _postprocess_action(action: Any, metadata: dict[str, Any]) -> Any:
...
+            # Display-ready output intentionally remains uint8 for ``np`` and
+            # ``pt`` so serving does not widen it only to encode it as bytes.
+            if output_type == "pt":
+                processed_video = video
+            elif output_type == "np":
+                processed_video = video.detach().cpu().numpy()
+            elif output_type == "pil":
+                vae_video = video.detach().cpu().permute(0, 4, 1, 2, 3).float().div_(127.5).sub_(1.0)
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py]: This allocates a full float32 [B,C,T,H,W] (~4.18 GB for 1280×720×189 frames) on CPU, negating the memory win for the pil path. The affine is also done twice (once to make uint8, once to undo it). Consider a direct uint8→PIL path via Image.fromarray per frame.

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py (hunk 1 above)
Comment (rahul-steiger-nv) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:836]: updated

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py (hunk 2 above)
Comment (rahul-steiger-nv) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py]: good point, I updated it
